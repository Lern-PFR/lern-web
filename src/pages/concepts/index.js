import ConceptDetailsPage from './ConceptDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ConceptDetailsPage,
};
