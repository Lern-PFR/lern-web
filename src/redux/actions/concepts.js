import * as ConceptsApi from 'api/conceptsApi';
import i18next from 'i18next';
import { toast } from 'react-toastify';
import { fetchSubject } from './subjects';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the concepts redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_CONCEPT_REQUEST: '@CONCEPTS/FETCH_REQUEST',
	FETCH_CONCEPT_SUCCESS: '@CONCEPTS/FETCH_SUCCESS',
	FETCH_CONCEPT_FAILURE: '@CONCEPTS/FETCH_FAILURE',

	FETCH_CONCEPT_LIST_REQUEST: '@CONCEPTS/FETCH_LIST_REQUEST',
	FETCH_CONCEPT_LIST_SUCCESS: '@CONCEPTS/FETCH_LIST_SUCCESS',
	FETCH_CONCEPT_LIST_FAILURE: '@CONCEPTS/FETCH_LIST_FAILURE',

	CREATE_CONCEPT_REQUEST: '@CONCEPTS/CREATE_REQUEST',
	CREATE_CONCEPT_SUCCESS: '@CONCEPTS/CREATE_SUCCESS',
	CREATE_CONCEPT_FAILURE: '@CONCEPTS/CREATE_FAILURE',

	UPDATE_CONCEPT_REQUEST: '@CONCEPTS/UPDATE_REQUEST',
	UPDATE_CONCEPT_SUCCESS: '@CONCEPTS/UPDATE_SUCCESS',
	UPDATE_CONCEPT_FAILURE: '@CONCEPTS/UPDATE_FAILURE',

	DELETE_CONCEPT_REQUEST: '@CONCEPTS/DELETE_REQUEST',
	DELETE_CONCEPT_SUCCESS: '@CONCEPTS/DELETE_SUCCESS',
	DELETE_CONCEPT_FAILURE: '@CONCEPTS/DELETE_FAILURE',

	CLEAR_CONCEPT_LIST: '@CONCEPTS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific concept fetching actions ///////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptRequest
 * @description Action triggered anytime a concept fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_REQUEST });

/**
 * @function
 * @name fetchConceptSuccess
 * @description Action triggered as a result to a successful concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The concept retrieved from the API.
 *
 * @returns {object}
 */
const fetchConceptSuccess = ({ concept }) => ({
	type: ActionTypes.FETCH_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name fetchConceptFailure
 * @description Action triggered as a result to a failed concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptListRequest
 * @description Action triggered anytime a concept list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptListRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_LIST_REQUEST });

/**
 * @function
 * @name fetchConceptListSuccess
 * @description Action triggered as a result to a successful concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	concepts	The list of concepts retrieved from the API.
 * @param {number}	totalCount	The total amount of concepts available for the module.
 *
 * @returns {object}
 */
const fetchConceptListSuccess = ({ concepts, totalCount }) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_SUCCESS,
	payload: { concepts, totalCount },
});

/**
 * @function
 * @name fetchConceptListFailure
 * @description Action triggered as a result to a failed concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptListFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createConceptRequest
 * @description Action triggered anytime a concept creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createConceptRequest = () => ({ type: ActionTypes.CREATE_CONCEPT_REQUEST });

/**
 * @function
 * @name createConceptSuccess
 * @description Action triggered as a result to a successfult concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The newly created concept.
 *
 * @returns {object}
 */
const createConceptSuccess = ({ concept }) => ({
	type: ActionTypes.CREATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name createConceptFailure
 * @description Action triggered as a result to a failed concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const createConceptFailure = (error) => ({
	type: ActionTypes.CREATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateConceptRequest
 * @description Action triggered anytime a concept edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateConceptRequest = () => ({ type: ActionTypes.UPDATE_CONCEPT_REQUEST });

/**
 * @function
 * @name updateConceptSuccess
 * @description Action triggered as a result to a successful concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The updated concept object.
 *
 * @returns {object}
 */
const updateConceptSuccess = ({ concept }) => ({
	type: ActionTypes.UPDATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name updateConceptFailure
 * @description Action triggered as a result to a failed concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const updateConceptFailure = (error) => ({
	type: ActionTypes.UPDATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept deletion actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteConceptRequest
 * @description Action triggered anytime a concept deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteConceptRequest = () => ({ type: ActionTypes.DELETE_CONCEPT_REQUEST });

/**
 * @function
 * @name deleteConceptSuccess
 * @description Action triggered as a result to a successful concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The deleted concept object.
 *
 * @returns {object}
 */
const deleteConceptSuccess = ({ concept }) => ({
	type: ActionTypes.DELETE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name deleteConceptFailure
 * @description Action triggered as a result to a failed concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteConceptFailure = (error) => ({
	type: ActionTypes.DELETE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list clearing actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name clearConceptListAction
 * @description Action triggered anytime the concept list is not needed anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const clearConceptListAction = () => ({ type: ActionTypes.CLEAR_CONCEPT_LIST });

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConcept
 * @description Method used to fetch the latest version of a specified concept.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 */
export const fetchConcept = (conceptId) => (dispatch) => {
	dispatch(fetchConceptRequest());

	return ConceptsApi.fetchConceptById(conceptId)
		.then((concept) => dispatch(fetchConceptSuccess({ concept })))
		.catch((error) => dispatch(fetchConceptFailure(error)));
};

/**
 * @function
 * @name fetchConceptListByModuleId
 * @description Method used to fetch all concepts from a specific module, identified by the moduleId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve concepts from.
 */
export const fetchConceptListByModuleId = (moduleId) => (dispatch) => {
	dispatch(fetchConceptListRequest());

	return ConceptsApi.fetchConceptsByModuleId(moduleId)
		.then(({ concepts, totalCount }) => dispatch(fetchConceptListSuccess({ concepts, totalCount })))
		.catch((error) => dispatch(fetchConceptListFailure(error)));
};

/**
 * @function
 * @name createConcept
 * @description Method used to create a new concept in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData The data to create the new concept from.
 */
export const createConcept = (conceptData) => (dispatch) => {
	dispatch(createConceptRequest());

	return ConceptsApi.createConcept(conceptData)
		.then(({ concept }) => dispatch(createConceptSuccess({ concept })))
		.catch((error) => dispatch(createConceptFailure(error)));
};

/**
 * @function
 * @name updateConcept
 * @description Method used to update an existing concept instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData	The data to update the concept with.
 * @param {string} conceptId	The id to identify the concept to update.
 */
export const updateConcept = (conceptData, conceptId) => (dispatch) => {
	dispatch(updateConceptRequest());

	return ConceptsApi.updateConcept(conceptData, conceptId)
		.then(({ concept }) => dispatch(updateConceptSuccess({ concept })))
		.catch((error) => dispatch(updateConceptFailure(error)));
};

/**
 * @function
 * @name deleteConcept
 * @description Method used to remove an existing concept from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId	The id identifying the concept to remove.
 * @param {string} subjectId	The id identifying the concept's parent subject.
 */
export const deleteConcept = (conceptId, subjectId) => (dispatch) => {
	dispatch(deleteConceptRequest());

	return ConceptsApi.deleteConcept(conceptId)
		.then((concept) => {
			dispatch(deleteConceptSuccess({ concept }));
			toast.success(i18next.t('concepts.deletion.toasts.success', { name: concept.title }));
			dispatch(fetchSubject(subjectId));
		})
		.catch((error) => dispatch(deleteConceptFailure(error)));
};

/**
 * @function
 * @name clearConceptList
 * @description Method used to clear the concept list stored in memory when the application does not need it anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Promise.resolve}
 */
export const clearConceptList = () => (dispatch) => {
	dispatch(clearConceptListAction());

	return Promise.resolve();
};
