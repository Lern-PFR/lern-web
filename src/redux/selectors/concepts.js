import { sortBy } from 'lodash';
import { createSelector } from 'reselect';

/**
 * @function
 * @name getConcepts
 * @description A selector callback which returns all concepts from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getConcepts = (state) => state?.concepts?.items ?? [];

/**
 * @function
 * @name getConceptById
 * @description A selector callback which returns the concept whose id matches the conceptId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getConceptById = createSelector(
	getConcepts,
	(_, conceptId) => conceptId,
	(concepts, conceptId) => {
		const concept = concepts.filter(({ id }) => (id === conceptId))?.[0] ?? undefined;

		if (concept === undefined) {
			return concept;
		}

		// @TODO: replace concept.courses with courses.lessons once the API has been updated.
		// Removes older versions of each lesson from the list.
		const lessons = Object.values((concept.courses || []).reduce((acc, currentLesson) => {
			const currentlyStoredLesson = acc[currentLesson.id];
			if (currentlyStoredLesson) {
				acc[currentLesson.id] = currentlyStoredLesson.version < currentLesson.version ? currentLesson : currentlyStoredLesson;
			} else {
				acc[currentLesson.id] = currentLesson;
			}

			return acc;
		}, {}));

		// @TODO: remove following snippet with the commented "return" snippet once the API has been updated.
		const { courses, ...result } = concept;

		return {
			...result,
			lessons: sortBy(lessons, 'order'),
		};

		// return {
		// 	...concept,
		//	lessons: sortBy(lessons),
		// };
	}
);

export {
	getConcepts,
	getConceptById,
};
