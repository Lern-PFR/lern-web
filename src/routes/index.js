import keys from './keys';

export {
	keys as default,
};
