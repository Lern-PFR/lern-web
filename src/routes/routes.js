/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/display-name */
import { PostSignupPage, SignInPage, SignupPage } from 'pages/auth';
import { ConceptCreationPage, ConceptDetailsPage, ConceptEditionPage } from 'pages/concepts';
import { SubjectCreationPage, SubjectDetailsPage, SubjectEditionPage, SubjectListPage, SubjectStructurePage } from 'pages/subjects';
import { ModuleCreationPage, ModuleDetailsPage, ModuleEditionPage } from 'pages/modules';
import { Homepage } from 'pages/home';
import { LessonCreationPage } from 'pages/lessons';
import keys from './keys';

/**
 * Routes to render
 * For authentication purposes, each route can have `allowAnonymous` and `allowAuthenticated` values
 * Default values are :
 *  - `allowAnonymous`		: false
 *  - `allowAuthenticated`	: true
 */
export default [
	{
		path: keys.app.default,
		exact: true,
		allowAnonymous: true,
		component: Homepage,
	},
	{
		path: keys.about.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: () => <p>About</p>,
	},
	{
		path: keys.contactUs.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: () => <p>Contact us</p>,
	},
	// Home
	{
		path: keys.home.default,
		exact: true,
		allowAnonymous: true,
		allowAuthenticated: true,
		component: Homepage,
	},
	// Auth
	{
		path: keys.auth.default,
		routes: [
			{
				path: keys.auth.login,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: SignInPage,
			},
			{
				path: keys.auth.logout,
				exact: true,
				component: () => <p>Logout</p>,
			},
			{
				path: keys.auth.forgottenPassword,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: () => <p>Forgot password</p>,
			},
			{
				path: keys.auth.signup,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: SignupPage,
			},
			{
				path: keys.auth.postSignup,
				allowAnonymous: true,
				allowAuthenticated: false,
				exact: true,
				component: PostSignupPage,
			},
		],
	},
	// Subjects
	{
		path: keys.subjects.default,
		routes: [
			{
				path: keys.subjects.list,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectListPage,
			},
			{
				path: keys.subjects.subjectStructure,
				allowAnonymous: true,
				allowAuthenticated: true,
				exact: true,
				component: SubjectStructurePage,
			},
			{
				path: keys.subjects.subjectCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectCreationPage,
			},
			{
				path: keys.subjects.subjectEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectEditionPage,
			},
			{
				path: keys.subjects.moduleCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleCreationPage,
			},
			{
				path: keys.subjects.subjectDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: SubjectDetailsPage,
			},
		],
	},
	// Modules
	{
		path: keys.modules.default,
		routes: [
			{
				path: keys.modules.moduleEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleEditionPage,
			},
			{
				path: keys.modules.conceptCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptCreationPage,
			},
			{
				path: keys.modules.moduleDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ModuleDetailsPage,
			},
		],
	},
	// Concepts
	{
		path: keys.concepts.default,
		routes: [
			{
				path: keys.concepts.conceptEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptEditionPage,
			},
			{
				path: keys.concepts.conceptDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptDetailsPage,
			},
			{
				path: keys.concepts.lessonCreation,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: LessonCreationPage,
			},
			{
				path: keys.concepts.lessonDetails,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: ConceptDetailsPage,
			},
		],
	},
	// Lessons
	{
		path: keys.lessons.default,
		routes: [
			{
				path: keys.lessons.lessonEdition,
				allowAnonymous: false,
				allowAuthenticated: true,
				exact: true,
				component: () => <p>Lesson edition</p>,
			},
		],
	},
];
