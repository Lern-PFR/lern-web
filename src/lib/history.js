import { createBrowserHistory } from 'history';
import { clearAlerts } from '../redux/actions/alerts';
import store from '../redux/store/configureStore';

const { dispatch } = store;

// Init applicationPathname
const history = createBrowserHistory();

history.listen(() => {
  dispatch(clearAlerts());
});

export default history;
