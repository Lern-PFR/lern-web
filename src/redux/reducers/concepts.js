import { ActionTypes } from 'redux/actions/concepts';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_CONCEPT_REQUEST:
		case ActionTypes.FETCH_CONCEPT_LIST_REQUEST:
		case ActionTypes.CREATE_CONCEPT_REQUEST:
		case ActionTypes.UPDATE_CONCEPT_REQUEST:
		case ActionTypes.DELETE_CONCEPT_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_CONCEPT_FAILURE:
		case ActionTypes.FETCH_CONCEPT_LIST_FAILURE:
		case ActionTypes.CREATE_CONCEPT_FAILURE:
		case ActionTypes.UPDATE_CONCEPT_FAILURE:
		case ActionTypes.DELETE_CONCEPT_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_CONCEPT_SUCCESS:
		case ActionTypes.DELETE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_CONCEPT_SUCCESS:
		case ActionTypes.UPDATE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.concept],
				totalCount: 1,
			};
		case ActionTypes.FETCH_CONCEPT_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.concepts],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_CONCEPT_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
