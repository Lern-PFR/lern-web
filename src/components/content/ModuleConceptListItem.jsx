import { useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { generatePath } from 'react-router';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

import routes from 'routes';
import { history } from 'routes/components/RouterProvider';

import { deleteConcept } from 'redux/actions/concepts';
import { StyledDiv, StyledListItem } from 'components/shared/styledElements';
import { BodyCopy, Brevier } from 'components/shared/typography';
import { OutlinedButton, OutlinedLinkButton } from 'components/shared/buttons';
import Link from 'components/shared/navigation/Link';

import { innerContentListItem, innerContentListItemActionsRow, moduleCard } from 'theme/contentEditionCommon/innerContentListTheme';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
		'&, & *': {
			cursor: 'pointer',
		},
	}
);

/**
 * @name ModuleConceptListItem
 * @description A component used to display a module's concept in the subject edition page.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {string} description	The current concept's description.
 * @param {string} id			The current concept's id.
 * @param {string} subjectId	The current concept's parent subject's id. Used for data refresh after deletion.
 * @param {string} title		THe current concept's title.
 */
const ModuleConceptListItem = ({ description, id, subjectId, title }) => {
	const dispatch = useDispatch();
	const { t } = useTranslation();

	const editPageRedirectionLink = useMemo(() => generatePath(routes.concepts.conceptEdition, { conceptId: id }), [id]);

	const onCardClick = useCallback(() => {
		history.push(editPageRedirectionLink);
	}, [editPageRedirectionLink]);

	const onDeleteBtnClick = useCallback((e) => {
		if (e?.stopPropagation) {
			e.stopPropagation();
		}

		dispatch(deleteConcept(id, subjectId));
	}, [dispatch, id, subjectId]);

	return (
		<ListCard key={id} onClick={onCardClick} {...moduleCard}>
			<BodyCopy margin="0">{title}</BodyCopy>
			<Brevier margin="0">{description}</Brevier>
			<StyledDiv {...innerContentListItemActionsRow}>
				<OutlinedLinkButton>
					<Link to={editPageRedirectionLink}>{t('modules.edition.concepts_list.links.edit')}</Link>
				</OutlinedLinkButton>
				<OutlinedButton onClick={onDeleteBtnClick}>
					{t('modules.edition.concepts_list.links.delete')}
				</OutlinedButton>
			</StyledDiv>
		</ListCard>
	);
};

ModuleConceptListItem.propTypes = {
	description: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	subjectId: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
};

export default ModuleConceptListItem;
