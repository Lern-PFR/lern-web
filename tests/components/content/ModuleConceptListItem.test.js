import { shallow } from 'enzyme';
import { act, fireEvent, render, screen } from '@testing-library/react';

import { Router, StaticRouter } from 'react-router';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import ModuleConceptListItem from 'components/content/ModuleConceptListItem';

import { ActionTypes } from 'redux/actions/concepts';
import { createMemoryHistory } from 'history';
import routes from 'routes';

describe('Module edition page - concept list item', () => {
	const mockStore = configureMockStore([thunk]);
	const store = mockStore({});
	const mockedConcept = { id: 'dummy_concept_id_1', title: 'dummy_concept_title_1', description: 'dummy_concept_desc_1' };

	afterEach(() => {
		jest.restoreAllMocks();
		jest.resetAllMocks();
	});

	describe('snapshots', () => {
		it('should match previous snapshot', () => {
			const sut = (
				<Provider store={store}>
					<ModuleConceptListItem {...mockedConcept} subjectId="1" />
				</Provider>
			);
			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('onClick events', () => {
		it('should dispatch a concept deletion request action upon click on the concept card\'s delete button.', () => {
			render(
				<StaticRouter>
					<Provider store={store}>
						<ModuleConceptListItem {...mockedConcept} subjectId="1" />
					</Provider>
				</StaticRouter>
			);

			act(() => {
				fireEvent.click(screen.getByText('modules.edition.concepts_list.links.delete'));
			});

			expect(store.getActions()).toEqual([{ type: ActionTypes.DELETE_CONCEPT_REQUEST }]);
		});

		it('should redirect to the concept\'s edition page url upon click on the card\'s edit button.', () => {
			const history = createMemoryHistory();

			render(
				<Router history={history}>
					<Provider store={store}>
						<ModuleConceptListItem {...mockedConcept} subjectId="1" />
					</Provider>
				</Router>
			);

			act(() => {
				fireEvent.click(screen.getByText('modules.edition.concepts_list.links.edit'));
			});

			expect(history.location.pathname).toEqual(routes.concepts.conceptEdition.replace(':conceptId', mockedConcept.id));
		});
	});
});
