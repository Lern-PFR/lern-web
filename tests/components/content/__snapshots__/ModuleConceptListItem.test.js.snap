// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module edition page - concept list item snapshots should match previous snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <ModuleConceptListItem
    description="dummy_concept_desc_1"
    id="dummy_concept_id_1"
    subjectId="1"
    title="dummy_concept_title_1"
  />
</ContextProvider>
`;
